pageserver/src/tenant/timeline.rs

```diff
@@ -2947,6 +2967,43 @@ impl Timeline {
             },
         )?;
 
+        // Determine N largest holes where N is number of compacted layers
+        let max_holes = deltas_to_compact.len();
+        let last_record_lsn = self.get_last_record_lsn();
+        let layers = self.layers.read().unwrap(); // Is'n it better to hold original layers lock till here?
+        let min_hole_range = (target_file_size / page_cache::PAGE_SZ as u64) as i128;
+        let min_hole_coverage_size = 3; // TODO: something more flexible?
+
+        // min-heap (reserve space for one more element added before eviction)
+        let mut heap: BinaryHeap<Hole> = BinaryHeap::with_capacity(max_holes + 1);
+        let mut prev: Option<Key> = None;
+        for (next_key, _next_lsn, _size) in itertools::process_results(
+            deltas_to_compact.iter().map(|l| l.key_iter(ctx)),
+            |iter_iter| iter_iter.kmerge_by(|a, b| a.0 <= b.0),
+        )? {
+            if let Some(prev_key) = prev {
+                // just first fast filter
+                if next_key.to_i128() - prev_key.to_i128() >= min_hole_range {
+                    let key_range = prev_key..next_key;
+                    let coverage_size = layers.image_coverage(&key_range, last_record_lsn)?.len();
+                    if coverage_size >= min_hole_coverage_size {
```

Comment:
I don't follow, why would there be image coverage on a hole with this patch?

Perhaps this is migration code from layermaps before this patch to versions with this patch?

Either way, the aim of filtering like this could be explained in a comment and/or commit message. Also note, the definition of a hole has also changed to be only about the image coverage not the key range.

Reply:
I added a comment explaining reason of using image coverage for measuring holes

Reply:
Thanks for the comment.

I still don't understand how we would get image layers covering these holes. Went through the compactions repartitioning and image layer creation but still am not seeing it.

Reply:
Okay, so it was the partitioning working together with this. I think in your picture you mean the second delta layer not to be indented. Though unsure of the second picture altogether, because it's the same as the first one.

How does it look like after this patch, or does it remain the same?

Reply:
The pictures above illustrate situation before this PR.
This PR will help to avoid such sparse delta layers.
So picture should be something like this:
```
<---------------><-----------------><-------------------->
<=======>                                  <====  ==  ==>
      <=  ==>                                                   <====>
<=>                                              <===>
<--------------->                         <-------------------->
```

Reply:
Linked this thread in the description, I think this is resolved.